source/opt/inst_debug_printf_pass.cpp

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2018 The Khronos Group Inc.
+// Copyright (c) 2018 Valve Corporation
+// Copyright (c) 2018 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inst_debug_printf_pass.h"
+
+#include "spirv/unified1/NonSemanticDebugPrintf.h"
+
+namespace spvtools {
+namespace opt {
+
+void InstDebugPrintfPass::GenOutputValues(Instruction* val_inst,
+                                          std::vector<uint32_t>* val_ids,
+                                          InstructionBuilder* builder) {
+  uint32_t val_ty_id = val_inst->type_id();
+  analysis::TypeManager* type_mgr = context()->get_type_mgr();
+  analysis::Type* val_ty = type_mgr->GetType(val_ty_id);
+  analysis::Vector* v_ty = val_ty->AsVector();
+  if (v_ty != nullptr) {
+    const analysis::Type* c_ty = v_ty->element_type();
+    uint32_t c_ty_id = type_mgr->GetId(c_ty);
+    for (uint32_t c = 0; c < v_ty->element_count(); ++c) {
+      Instruction* c_inst = builder->AddIdLiteralOp(
+          c_ty_id, SpvOpCompositeExtract, val_inst->result_id(), c);
+      GenOutputValues(c_inst, val_ids, builder);
+    }
+    return;
+  }
+  analysis::Bool* b_ty = val_ty->AsBool();
+  if (b_ty != nullptr) {
```

Comment:
I think you can do a switch for this using `val_ty->Kind()`.  This would make it a little cleaner.

Reply:
k